### PR TITLE
scheduler: redesigned `progressiveRank` of the hot-region-scheduler

### PR DESCRIPTION
**Pull request by HunDunDM**
### What problem does this PR solve?

Issue Number: Ref #4949

The score is expanded from -3 ~ 0 to -4 ~ 1, and the previous design problem that the score of the first dimension is better and the second dimension is not worsened is -1, and the score of the first dimension is not worsened and the second dimension is better is -2. The solution still needs to take a score of -1 and below. The original 0 points become 0 and 1.

### Check List

Tests

- Unit test

Code changes

Side effects

Related changes

### Release note

```release-note
None
```

### CODE REVIEW

File: server/schedulers/hot_region.go [hunk 1]
```diff
@@ -1029,12 +1037,12 @@ func (bs *balanceSolver) buildOperators() (ops []*operator.Operator, infl *stati
 	targetLabel := strconv.FormatUint(dstStoreID, 10)
 	dim := ""
 	switch bs.cur.progressiveRank {
-	case -3:
+	case -4:
 		dim = "all"
+	case -3, -1:
```
Comment (lhy1024) [inline server/schedulers/hot_region.go:1042]: Do we need to distinguish this two cases in metrics?

File: server/schedulers/hot_region.go [hunk 2]
```diff
@@ -882,37 +895,32 @@ func (bs *balanceSolver) betterThan(old *solution) bool {
...
-					return secondCmp > 0
-				}
-				fallthrough
-			case -1: // firstPriorityDecRatio <= greatDecRatio
-				if firstCmp != 0 {
-					// prefer region with larger first priority rate, to converge faster
-					return firstCmp > 0
-				}
+			return secondCmp > 0
+		case -3, -1:
```
Comment (lhy1024) [inline server/schedulers/hot_region.go:910]: I think we need to split these two cases. -3 should be more similar with -2

Comment (HunDunDM) [reply server/schedulers/hot_region.go:910]: I wrote them separately at first, but the two seem to have the same processing logic at the moment.

File: server/schedulers/hot_region.go [hunk 3]
```diff
@@ -778,41 +778,46 @@ func (bs *balanceSolver) checkDstByPriorityAndTolerance(maxLoad, expect *statist
...
 
 	isFirstBetter, isSecondBetter := bs.isBetter(bs.firstPriority), bs.isBetter(bs.secondPriority)
-	isFirstNotWorsened := !isFirstBetter && bs.isNotWorsened(bs.firstPriority)
-	isSecondNotWorsened := !isSecondBetter && bs.isNotWorsened(bs.secondPriority)
+	isFirstNotWorsened := isFirstBetter || bs.isNotWorsened(bs.firstPriority)
+	isSecondNotWorsened := isSecondBetter || bs.isNotWorsened(bs.secondPriority)
 	switch {
 	case isFirstBetter && isSecondBetter:
-		// If belong to the case, two dim will be more balanced, the best choice.
+		// If belonging to the case, all two dim will be more balanced, the best choice.
```
Comment (lhy1024) [inline server/schedulers/hot_region.go:801]:
```suggestion
		// If belonging to the case, both two dim will be more balanced, the best choice.
```

Comment (lhy1024) [inline server/schedulers/hot_region.go:801]:
```suggestion
		// If belonging to the case, both two dim will be more balanced, the best choice.
```

File: server/schedulers/hot_region.go [hunk 4]
```diff
@@ -876,37 +900,32 @@ func (bs *balanceSolver) betterThan(old *solution) bool {
...
 			kind := statistics.GetRegionStatKind(statistics.Write, bs.firstPriority)
-			switch {
-			case bs.cur.srcPeerStat.GetLoad(kind) > old.srcPeerStat.GetLoad(kind):
-				return true
-			case bs.cur.srcPeerStat.GetLoad(kind) < old.srcPeerStat.GetLoad(kind):
-				return false
+			return bs.cur.srcPeerStat.GetLoad(kind) > old.srcPeerStat.GetLoad(kind)
+		}
+
+		firstCmp, secondCmp := bs.getRkCmpPriorities(old)
```
Comment (lhy1024) [inline server/schedulers/hot_region.go:908]: need to add comment `We will firstly consider ensuring converge faster, secondly reduce oscillation`

File: server/schedulers/hot_region.go (hunk 2 above)
Comment (lhy1024) [reply server/schedulers/hot_region.go:915]: I think it will be friendly to read code and modify if split these cases.

**Review by nolouch — APPROVED** [review]
lgtm

**Review by lhy1024 — APPROVED** [review]

**Pull request merged**